servers/physics_2d/space_2d_sw.cpp

```diff
@@ -638,17 +641,23 @@ int Space2DSW::test_body_ray_separation(Body2DSW *p_body, const Transform2D &p_t
 						if (ray_index != -1) {
 							PhysicsServer2D::SeparationResult &result = r_results[ray_index];
 
+							recover_count += cbk.amount;
 							for (int k = 0; k < cbk.amount; k++) {
 								Vector2 a = sr[k * 2 + 0];
 								Vector2 b = sr[k * 2 + 1];
+								Vector2 separation = (b - a);
 
-								recover_motion += (b - a) / cbk.amount;
+								// Apply recovery without margin.
+								float depth = separation.length();
+								float separation_depth = depth - p_margin;
+								if (separation_depth > 0.0) {
+									recover_motion += separation * (separation_depth / depth);
+								}
 
-								float depth = a.distance_to(b);
 								if (depth > result.collision_depth) {
 									result.collision_depth = depth;
 									result.collision_point = b;
-									result.collision_normal = (b - a).normalized();
+									result.collision_normal = ray_normal;
```

Comment:
Always using raycast normal as a result, to make sure it only does separation, as described in this article:
https://godotengine.org/article/godot-31-will-get-many-improvements-kinematicbody